### PR TITLE
add x-chainweb-time header

### PR DESCRIPTION
**Pull request by larskuhtz**
This adds a custom HTTP response header to all API endpoints that contains the POSIX timestamp of server.

Clients can check that and decide whether they (or the server) is out of sync. Clients that can't reliably sync there clock can use this header to successfully submit requests.

### CODE REVIEW

**Comment by larskuhtz** [comment]
Another way to implement this would have been to add the header to the `ChainwebEndpointApi` type. That would cause the header to show up in servant generated documentation, but it would also require to change all handler functions in the code.

File: src/Chainweb/RestAPI.hs [hunk 1]
```diff
@@ -224,15 +228,27 @@ chainwebApplication
...
     { corsRequestHeaders = simpleHeaders
     }
 
+chainwebTime :: Middleware
+chainwebTime app req resp = app req resp'
+  where
+    resp' res = do
+        timestamp <- realToFrac <$> getPOSIXTime :: IO Double
+        resp $ mapResponseHeaders
+            ((:) ("X-Chainweb-Time", B8.pack (printf "%.0f" timestamp)))
```
Comment (gregorycollins) [inline src/Chainweb/RestAPI.hs]: a) avoid `Double` for timestamps -- we already have a dep on the clock package so here I would prefer http://hackage.haskell.org/package/clock-0.8/docs/src/System.Clock.html#getTime with CLOCK_REALTIME_COARSE -- easy to convert the resulting `TimeSpec` to `Ratio Int64` for `printf`, or do the job by hand with `div` on the `nsec` field of the timespec

b) the header should be called "X-Server-Timestamp" (and the bike shed should be blue)

Comment (larskuhtz) [reply src/Chainweb/RestAPI.hs]: It seems that `CLOCK_REALTIME_COARSE` is available only on linux (or at least it isn't obvious from the package code if it is also available on other platforms).

There's also no code for easy detection or graceful fallback available. So, I'll stick with `Realtime`.

I don't think that the `RealTime` clock is (much) faster than `getPOSIXTime` from `time` and we already use `time` in the same code path. But happy to use clock, instead.

File: src/Servant/Client_.hs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    ( Functor, Applicative, Monad, MonadIO
+    , Generic
+    , MonadError ClientError, MonadThrow , MonadCatch
+    )
+
+instance MonadReader ClientEnv ClientM_ where
+    ask = ClientM_ $ lift ask
+    local f = ClientM_ . mapReaderT (local f) . unClientM_
+    {-# INLINE ask #-}
+    {-# INLINE local #-}
```
Comment (emilypi) [inline src/Servant/Client_.hs]: isn't this the derived instance for `ClientM_`?

File: src/Servant/Client_.hs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+-- | Locally overwrite 'Request's made the inner computation.
+--
+modifyRequest :: (Request -> IO Request) -> ClientM_ a -> ClientM_ a
+modifyRequest f = ClientM_ . local (set modReq f) . unClientM_
+{-# INLINE modifyResponse #-}
+
+-- | Locally overwrite 'Response's received the inner computation.
+--
+modifyResponse :: (Response -> IO Response) -> ClientM_ a -> ClientM_ a
+modifyResponse f = ClientM_ . local (set modRes f) . unClientM_
```
Comment (emilypi) [inline src/Servant/Client_.hs]: same as above - isn't this just `local (set modRes f)` for `ClientM_`?

File: test/Chainweb/Test/RestAPI.hs [hunk 4]
```diff
@@ -186,6 +229,103 @@ simpleClientSession envIO cid =
                 (Expected h)
                 (Actual r)
 
+        -- brancheHeaders
```
Comment (emilypi) [inline test/Chainweb/Test/RestAPI.hs]: typo

File: test/Chainweb/Test/RestAPI.hs [hunk 5]
```diff
@@ -186,6 +229,103 @@ simpleClientSession envIO cid =
...
+
+        forM_ (newHeaders `zip` drop 2 newHeaders) $ \(h0, h1) -> do
+            void $ liftIO $ step "branchHeadersClient: get two block headers with lower and upper bound"
+            bhs5 <- branchHeadersClient version cid Nothing Nothing Nothing Nothing
+                (BranchBounds (HS.singleton (LowerBound $ key h0)) (HS.singleton (UpperBound $ key h1)))
+            assertExpectation "branchHeadersClient returned wrong number of entries"
+                (Expected 2)
+                (Actual $ _pageLimit bhs5)
+
+        -- brancheHeaders
```
Comment (emilypi) [inline test/Chainweb/Test/RestAPI.hs]: (see above)

File: test/Chainweb/Test/RestAPI.hs (hunk 4 above)
Comment (larskuhtz) [inline test/Chainweb/Test/RestAPI.hs]:
```suggestion
        -- branchHeaders
```

File: test/Chainweb/Test/RestAPI.hs (hunk 5 above)
Comment (larskuhtz) [inline test/Chainweb/Test/RestAPI.hs]:
```suggestion
        -- branchHeaders
```

File: src/Servant/Client_.hs (hunk 2 above)
Comment (larskuhtz) [reply src/Servant/Client_.hs]: `ClientM_` implements `MonadReader ClientEnv`, i.e. it exposes the environment of the underlying `ClientM` monad. 

The fact that `ClientEnv_` itself is a `ReaderT Mods` is a hidden implementation detail and not exported.

File: src/Servant/Client_.hs (hunk 3 above)
Comment (larskuhtz) [reply src/Servant/Client_.hs]: `ClientM_` implements `MonadReader ClientEnv`. In order to access the `Mods` environment, we have to unwrap it to get the internal `ReaderT Mods`.

**Pull request merged**